src/main/java/org/jenkinsci/plugins/workflow/graphanalysis/AbstractFlowScanner.java

```diff
@@ -0,0 +1,255 @@
+ */
+public abstract class AbstractFlowScanner implements Iterable <FlowNode>, Filterator<FlowNode> {
+
+    protected FlowNode _current;
+
+    protected FlowNode _next;
+
+    protected Collection<FlowNode> _blackList = Collections.EMPTY_SET;
+
+    /** Convert stop nodes to a collection that can efficiently be checked for membership, handling null if needed */
+    @Nonnull
+    protected Collection<FlowNode> convertToFastCheckable(@CheckForNull Collection<FlowNode> nodeCollection) {
+        if (nodeCollection == null || nodeCollection.size()==0) {
+            return Collections.EMPTY_SET;
+        } else  if (nodeCollection.size() == 1) {
+            return Collections.singleton(nodeCollection.iterator().next());
+        } else if (nodeCollection instanceof Set) {
+            return nodeCollection;
+        }
+        return nodeCollection.size() > 5 ? new HashSet<FlowNode>(nodeCollection) : nodeCollection;
+    }
+
+    /**
+     * Set up for iteration/analysis on a graph of nodes, initializing the internal state
+     * @param heads The head nodes we start walking from (the most recently executed nodes,
+     *              i.e. FlowExecution.getCurrentHeads()
+     * @param blackList Nodes that we cannot visit or walk past (useful to limit scanning to only nodes after a specific point)
+     * @return True if we can have nodes to work with, otherwise false
+     */
+    public boolean setup(@CheckForNull Collection<FlowNode> heads, @CheckForNull Collection<FlowNode> blackList) {
+        if (heads == null || heads.size() == 0) {
+            return false;
+        }
+        Collection<FlowNode> fastEndNodes = convertToFastCheckable(blackList);
+        HashSet<FlowNode> filteredHeads = new HashSet<FlowNode>(heads);
+        filteredHeads.removeAll(fastEndNodes);
+
+        if (filteredHeads.size() == 0) {
+            return false;
+        }
+
+        reset();
+        _blackList = fastEndNodes;
+        setHeads(filteredHeads);
+        return true;
+    }
+
+    /**
+     * Set up for iteration/analysis on a graph of nodes, initializing the internal state
+     * @param head The head FlowNode to start walking back from
+     * @param blackList Nodes that we cannot visit or walk past (useful to limit scanning to only nodes after a specific point)
+     *                  null or empty collection means none
+     * @return True if we can have nodes to work with, otherwise false
+     */
+    public boolean setup(@CheckForNull FlowNode head, @CheckForNull Collection<FlowNode> blackList) {
+        if (head == null) {
+            return false;
+        }
+        return setup(Collections.singleton(head), blackList);
+    }
+
+    public boolean setup(@CheckForNull FlowNode head) {
+        if (head == null) {
+            return false;
+        }
+        return setup(Collections.singleton(head), Collections.EMPTY_SET);
+    }
+
+    /** Public APIs need to invoke this before searches */
+    protected abstract void reset();
```

Comment:
> Is this stateful API design really necessary?

Yes.  The only logic that doesn't need state is trivial, and we'd need to create a ton of throwaway Iterators otherwise.